Sketch on the map sample

Rachael-E opened the pull request:
Sketch on map implementation. The sample shows how to sketch directly on to a map view. Ready for review @tschie when you have a moment, thanks!

Rachael-E committed:
Create SketchOnMapSample.java

Initialising sketch on map sample

Rachael-E committed:
Update SketchOnMapSample.java

Adding combobox for graphic selection

Rachael-E committed:
Update SketchOnMapSample.java

added buttons for controls

Rachael-E committed:
Update SketchOnMapSample.java

Configuring edit button to activate when a graphic is selected from the graphic overlay and deactivated when mouse clicks away.

Rachael-E committed:
Update SketchOnMapSample.java

Can now update a graphic with new graphic in the edit button without creating multiple new graphics in its place.

Rachael-E committed:
Update SketchOnMapSample.java

Added a cancel sketch button for when user wants to exit sketch mode. This activates when geometry is changed.

Rachael-E committed:
Update SketchOnMapSample.java

If sketch editor is capable of enabling redo or undo, enable those buttons. Otherwise, disable them.

Rachael-E committed:
Update SketchOnMapSample.java

changes to redo and undo

Rachael-E committed:
Update SketchOnMapSample.java

Cancel button simply stops the sketch editor, allows the option to select a graphic, clears the selection and sets the cancel and undo buttons to disabled.

Rachael-E committed:
Update SketchOnMapSample.java

Tidied up old logic from cancel button

Rachael-E committed:
Update SketchOnMapSample.java

Added relevant information to appear on save button depending on mode selected

Rachael-E committed:
Update SketchOnMapSample.java

When a new sketch is being drawn, check to see if the sketch editor geometry is valid and if so, enable to save to graphics overlay button.

Rachael-E committed:
Update SketchOnMapSample.java

When selecting a new sketch, check to see if there is already a graphic in the graphics layer and if so, enable the clear graphics overlay button.

Rachael-E committed:
Update SketchOnMapSample.java

Tidying up the edit button functionality and adding comments.

Rachael-E committed:
Update SketchOnMapSample.java

Code tidy through out

Rachael-E committed:
Update SketchOnMapSample.java

Re-ordered ArcGIS runtime imports

Rachael-E committed:
Update SketchOnMapSample.java

When a user draws a graphic and saves it to the graphics overlay, they can now continue drawing it and have the option to exit the sketch editor.

Rachael-E committed:
Update SketchOnMapSample.java

Allow button text to change depending on user interaction

Rachael-E committed:
Initialised ReadMe and code edits

Initialised ReadMe and tidying up of code.

Rachael-E committed:
Update README.md

Added Relevant API section to readme

Rachael-E committed:
Updates to readme, and code tidying

Updates to readme, and code tidying

Rachael-E committed:
Added image and code tidying

Added image and code tidying

Rachael-E committed:
Update README.md

Added path to image

Rachael-E committed:
Added new image and code edits

Code edits are removing label and using button prompt instead.

Rachael-E committed:
Update SketchOnMapSample.java

removed redundant reference to set border for button

Rachael-E committed:
Update SketchOnMapSample.java

Removed system print line

Rachael-E committed:
Update SketchOnMapSample.java

Adding additional if else statements to catch edge cases

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapSample.java:
```diff
@@ -0,0 +1,415 @@
+        // get the selected item from the sketchComboBox to determine which sketch style to use
+        SketchCreationMode sketchCreationMode = sketchComboBox.getSelectionModel().getSelectedItem();
+
+        try {
```
You don't need this try/catch and switch. Just use
```java
sketchEditor.start(sketchCreationMode);
```

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapSample.java:
```diff
@@ -0,0 +1,415 @@
+          saveButton.setText("Save sketch to graphics overlay");
+          saveButton.setDisable(false);
+          stopSketchButton.setDisable(false);
+        } else {saveButton.setDisable(true);}
```
should be on a new line

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapSample.java:
```diff
@@ -0,0 +1,415 @@
+      controlsVBox.getStyleClass().add("panel-region");
+
+      // create a flow pane for placing buttons side by side within the control box
+      FlowPane flowPaneUndoRedo = new FlowPane(Orientation.HORIZONTAL, 70, 10, undoButton, redoButton);
```
A FlowPane is really only needed when you want the layout to be responsive to layout width/height. Like if you wanted to change the layout going from a desktop sized window to a mobile sized window. I think here a simple VBox containing an HBox for the Undo/Redo buttons would suffice.

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapSample.java:
```diff
@@ -0,0 +1,415 @@
+        graphic.setGeometry(sketchGeometry);
+      } else {
+        graphic = new Graphic(sketchGeometry);
+        if (graphic.getGeometry().getGeometryType() == GeometryType.POLYGON) {
```
use a switch here
```java
switch (sketchGeometry.getGeometryType()) {
    case POLYGON:
        graphic.setSymbol(fillSymbol);
        break;
    case POLYLINE:
        graphic.setSymbol(lineSymbol);
        break;
    case POINT:
    case MULTIPOINT:
        graphic.setSymbol(pointSymbol);
}
```

tschie replied:
Another alternative is to have three graphics overlays: points, lines, and polygons. Then specify a renderer for the overlay with the appropriate symbol. The advantage is you can just create graphics with geometry and add it to the right overlay and it will inherit the symbol of the overlay's renderer.

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapSample.java:
```diff
@@ -0,0 +1,415 @@
+
+      // define symbols for graphics
+      pointSymbol = new SimpleMarkerSymbol(SimpleMarkerSymbol.Style.SQUARE, 0xFFFF0000, 20);
+      lineSymbol = new SimpleLineSymbol(SimpleLineSymbol.Style.SOLID, 0xFF64c113, 4);
```
A comment above each symbol describing the color would be helpful: "thin green line"

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/README.md:
```diff
@@ -0,0 +1,47 @@
+
+<p>Use the Sketch Editor to edit, or sketch a new point, line, or polygon geometry on to a map. </p>
+
+<p><img src="SketchOnMap.png"/></p>
```
Very cool picture!

Rachael-E replied:
Thanks! It's a good tool to highlight the previous advances of the glacier :)

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapSample.java:
```diff
@@ -0,0 +1,415 @@
+      // create buttons for user interaction
+      undoButton = new Button("Undo sketch");
+      redoButton = new Button("Redo sketch");
+      clearButton = new Button("Clear graphics overlay");
```
I think we should use less API terminology: "Clear sketches", "Save sketch"

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapSample.java:
```diff
@@ -0,0 +1,415 @@
+      // disable all buttons when starting application
+      disableButtons();
+      // set prompt text on combobox when starting application
+      sketchComboBox.setPromptText("    -- Select a graphic to sketch --");
```
"Select sketch mode"

Rachael-E committed:
Update SketchOnMapSample.java

Simplified sketchEditor start method by removing switch case

Rachael-E committed:
Update SketchOnMapSample.java

Simplifying button descriptions

Rachael-E committed:
Added sketch icons to resources and changed UI

UI now consists of buttons to use for sketches rather than a drop down box.

Rachael-E committed:
Update SketchOnMapSample.java

Add comments to symbol types

Rachael-E committed:
Update SketchOnMapSample.java

Flowpane replaced with HBox for undo redo buttons

Rachael-E committed:
Initialising controller and fxml file

Initialising controller and fxml file

Rachael-E committed:
Setting up Controller for FXML

Setting up Controller for FXML

Rachael-E committed:
Refactoring for fxml structure

Added buttons to fxml and removed them from controller class

Rachael-E committed:
Refactoring on action buttons to methods

Refactoring on save, stop and edit actions to methods, for use in FXML

Rachael-E committed:
Remove unused imports

Remove unused imports

Rachael-E committed:
Delete SketchOnMapSample_Old.java

Remove previous version of self-contained sketch on map sample java class (prior to splitting into MVC)

Rachael-E committed:
Update SketchOnMap.png

Updated image to show new UI with sketch buttons

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapController.java:
```diff
@@ -0,0 +1,343 @@
+    fillSymbol = new SimpleFillSymbol(SimpleFillSymbol.Style.CROSS, 0x40FFA9A9,polygonLineSymbol);
+
+    // start sketch editor with relevant mode for each sketch button
+    createPointButton.setOnAction(event -> {
```
Make these `@FXML` methods like the others

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapController.java:
```diff
@@ -0,0 +1,343 @@
+
+
+    // disable all buttons except sketch buttons when starting application
+    disableButtons();
```
set `disable=true` on the buttons in the FXML to initialize

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapController.java:
```diff
@@ -0,0 +1,343 @@
+      }
+
+      // if the sketch editor can redo, enable the redo button otherwise disable it
+      if (sketchEditor.canRedo()) {
```
Could simplify to `redoButton.setDisable(sketchEditor.canRedo())`. Same for the other one.

You can remove the comment. The code is self-explanatory.

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapSample.java:
```diff
@@ -0,0 +1,96 @@
+
+package com.esri.samples.displayinformation.sketch_on_map;
+
+import com.esri.arcgisruntime.concurrent.ListenableFuture;
```
Unused imports

tschie commented on src/main/resources/fxml/sketch_on_map.fxml:
```diff
@@ -0,0 +1,87 @@
+        <HBox maxWidth="200" maxHeight="50" spacing="6">
+            <Button fx:id="createPointButton" style="-fx-graphic: url('/icons/pointSketch.png')">
+                <tooltip>
+                    <Tooltip text="Sketch single point"></Tooltip>
```
For empty tags, you can do `<Tooltip text="Sketch single point"/>`

Rachael-E committed:
Update SketchOnMapSample.java

Removed unused imports

Rachael-E committed:
Reorder imports and rename stop button

Rachael-E committed:
Updated icons and paths in FXML

Rachael-E committed:
Update SketchOnMapController.java

Made changes as suggested at Code Review to make sketch editor more UI intuitive

Rachael-E committed:
Refactor

Move an import statement and remove a setText reference for a button

Rachael-E committed:
Update README.md

Edits to readme, updating tags and description following UI changes

Rachael-E committed:
Update README.md

Including polyline in the description

Rachael-E committed:
Update SketchOnMap.png

Updated image to show new UI controls

Rachael-E committed:
Update SketchOnMapController.java

Tidying up of code to ensure every private function has a description, and removing white space.

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapController.java:
```diff
@@ -0,0 +1,315 @@
+  }
+
+  /**
+   * clear the graphics overlay of any saved graphics
```
Javadoc is a little more strict than inline comments. Should capitalize first letter and have proper punctuation.

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapController.java:
```diff
@@ -0,0 +1,315 @@
+
+    Geometry sketchGeometry = sketchEditor.getGeometry();
+
+    // if an existed graphic being edited: get the selected graphic, set its geometry to that of the sketch editor geometry
```
~~existed~~ existing

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapController.java:
```diff
@@ -0,0 +1,315 @@
+   */
+  @FXML
+  private void handleSaveButtonClicked() {
+    // if the sketch isn't valid, stop the sketch editor.
```
I'm not sure this can ever happen. See line 92.

tschie commented on src/main/java/com/esri/samples/displayinformation/sketch_on_map/SketchOnMapController.java:
```diff
@@ -0,0 +1,315 @@
+    disableButtons();
+    // set text to inform the user the sketch is disabled
+    stopButton.setDisable(false);
+    // allow graphics to be selected after stopS button is used.
```
typo stopS

tschie reviewed (approved):
Looks good! Just a few really minor things left

Rachael-E committed:
Update SketchOnMapController.java

Tidy up Javadoc

Rachael-E committed:
Update SketchOnMapController.java

Removed redundant code logic

Rachael-E committed:
Update SketchOnMapController.java

Created local variable in place of member variable
